lib/ecto/query/builder/join.ex

```diff
@@ -229,6 +231,26 @@ defmodule Ecto.Query.Builder.Join do
     end
   end
 
+  defp ensure_on(nil, join) do
+    unless join[:assoc] do
+      maybe_source =
+        with {source, alias} <- join[:source],
+          source when source != nil <- source || alias do
+          " on #{inspect(source)}"
+        else
+          _ -> ""
+        end
+
+      Logger.warning(
+        "#{join[:file]}:#{join[:line]} Missing `:on` in join#{maybe_source}, defaulting to `on: true`."
+      )
```

Comment:
I struggled a bit to come up with a useful warning with the proper line number, else it would be quite un-helpful:

<img width="616" alt="Screenshot 2023-01-07 at 9 31 58" src="https://user-images.githubusercontent.com/11598866/211123112-5e19fa61-cf55-4226-9d22-eca9e86c7a11.png">

With `IO.warn` (stacktrace isn't helpful):
<img width="597" alt="Screenshot 2023-01-07 at 9 55 50" src="https://user-images.githubusercontent.com/11598866/211123644-e97a4fdf-bf8f-4aa8-a5ad-599c7d5d4cd2.png">